### ISSUE TITLE
Accessibility - cannot navigate the tree via keyboard

### ISSUE THREAD

**Issue by phudson-he** [mbraak/jqTree#573]
The tree nodes cannot be navigated by using only keyboard. For accessibility purposes could the code perhaps be changed to add href="#" to any anchor tags, and include the ability to press the keyboard enter key to expand/collapse/action nodes etc?

**Comment by mbraak** [comment]
Jqtree has keyboard support by default. You can use the `right arrow` to open a node and `left arrow` to close one.

**Comment by phudson-he** [comment]
Unfortunately I cannot check that because I cannot tab to any nodes. If I manually add an href, say for example href="#" then I can at least tab to it using the keyboard. Just for info Im using Chrome (Version 76.0.3809.100).

**Comment by mbraak** [comment]
You can only use keys after you select a node using a mouse.

Maybe there is an alternative. I'll look into it.

**Comment by phudson-he** [comment]
Thanks, much appreciated. Yes that would be fine but for someone who can only use a keyboard, they wouldnt be able to get to any node.

Just wondering would it be as easy as adding an href attribute to /src/elements_renderer.ts - line 159, change...

buttonLink.setAttribute("role", "presentation");
buttonLink.setAttribute("aria-hidden", "true");

...to...

buttonLink.setAttribute("href", "#");
buttonLink.setAttribute("role", "presentation");
buttonLink.setAttribute("aria-hidden", "true");

Just a thought :)

**Comment by mbraak** [comment]
Thanks. I'll try the solution with the `href`.

**Comment by phudson-he** [comment]
Thats great. No pressure but I'd like a new release on here yesterday haha.

**Comment by mbraak** [comment]
A possible workaround is to select the first node. That will make the tree focusable using the keyboard.

```js
const $tree = $("#tree1");

$tree.on("tree.init", function() {
    const root = $tree.tree("getTree");

    if (root.hasChildren()) {
        $tree.tree("selectNode", root.children[0]);
    }
});

$tree.tree({
    data: ExampleData.exampleData
});
```

Hopefully this works. I will also look for a solution that works without this workaround.

**Comment by AJ-HistoricEngland** [comment]
@mbraak , I work with @phudson-he and will be keeping an eye on this thread while he is on leave.

Just worth noting for context that we are looking at this as it has failed an accessibility audit within our application, which is subject to new EU requirements for government websites. Our users must be able to navigate into and around this tree at any point in the user's workflow using keyboard only.

Therefore, giving focus to the root node at initialization won't work in most cases unfortunately.

Paul has identified that browsers will automatically tab into the first item (root?) of the tree structure if it has this href attribute set with at least a "#" value.

Thanks for looking into this!

**Comment by mbraak** [comment]
You can add the `href` to all buttons using the following customization code:

```js
const $tree = $("#tree1");

$tree.tree({
    data: ExampleData.exampleData,
    onCreateLi: function(node, $li) {
        var $button = $li.children(".jqtree-element").find("a.jqtree-toggler");
        $button.attr("href", "#");
    }
});
```

This seems to work. Unfortunately I cannot add this to jqtree by default because it conflicts with the existing keyboard functionality.

**Comment by mbraak** [comment]
This isssue has been open for a long time without any progress. Closing.
